### PR TITLE
Remove outpoint lookup when UTXO is spent

### PR DESCRIPTION
**Pull request by bokobza**
Saved from oblivion (Nov' 18).

Remove an outpoint from the lookup table when a UTXO is spent.
Recovered a 90MB wallet containing about 60,000 received outputs and 55,000 spent outputs items, all recovered without problem.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 1]
```diff
@@ -1340,6 +1348,20 @@ private void AddInputKeysLookupLocked(TransactionData transactionData)
...
+        /// Remove from the list of unspent outputs kept in memory.
+        /// </summary>
+        private void RemoveInputKeysLookupLock(TransactionData transactionData)
+        {
+            Guard.NotNull(transactionData, nameof(transactionData));
+            Guard.NotNull(transactionData.SpendingDetails, nameof(transactionData.SpendingDetails));
+
+            lock (this.lockObject)
+            {
+                this.outpointLookup.Remove(new OutPoint(transactionData.Id, transactionData.Index));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Is it a good idea to new up a new `OutPoint` in order to remove it?

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Is there perhaps a better way to do this?

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: That's the best way.

The `outpointLookup` object is a `Dictionary<OutPoint, TransactionData>` so you can only remove an item by specifying the key, here an `OutPoint`.
The OutPoint class overrides the `==` operator so comparison is serving our purpose here.
https://github.com/stratisproject/StratisBitcoinFullNode/blob/13b272602e63c33fee51ed0f5dc4dc29646a24b0/src/NBitcoin/Transaction.cs#L129-L140

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Yes the implicit operator will work but if you want to remove the item without a new instance of `Outpoint` you can do this:

```                
var toRemove = this.outpointLookup.FirstOrDefault(op => op.Key.Hash == transactionData.Id && op.Key.N == transactionData.Index);
this.outpointLookup.Remove(toRemove.Key);
```

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: true, but isn't `var toRemove` a new instance of `OutPoint`?

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: In this case `toRemove` will be the actual intance in the dictionary that we want to remove.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 2]
```diff
@@ -1297,7 +1303,9 @@ public void LoadKeysLookupLock()
                         if (address.Pubkey != null)
                             this.scriptToAddressLookup[address.Pubkey] = address;
 
-                        foreach (TransactionData transaction in address.Transactions)
+                        // Get the UTXOs that are unspent or spent but not confirmed.
+                        // We only exclude from the list the confirmed spent UTXOs.
+                        foreach (TransactionData transaction in address.Transactions.Where(t => t.SpendingDetails?.BlockHeight == null))
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: We call this again after a reorg right?

**Pull request merged**